// @canner/image-upload/lib/Container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _templateObject = _taggedTemplateLiteral(['\n  padding: 30px;\n'], ['\n  padding: 30px;\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _antd = require('antd');

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reactIntl = require('react-intl');

var _UploadImage = require('./UploadImage');

var _UploadImage2 = _interopRequireDefault(_UploadImage);

var _DefaultImage = require('./DefaultImage');

var _DefaultImage2 = _interopRequireDefault(_DefaultImage);

var _UrlImage = require('./UrlImage');

var _UrlImage2 = _interopRequireDefault(_UrlImage);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var TabPane = _antd.Tabs.TabPane;

var Content = _styledComponents2.default.div(_templateObject);

var Container = function (_Component) {
  _inherits(Container, _Component);

  function Container() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Container);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Container.__proto__ || Object.getPrototypeOf(Container)).call.apply(_ref, [this].concat(args))), _this), _this.finishEdit = function (e) {
      var closeEditPopup = _this.props.closeEditPopup;

      e.preventDefault();
      e.stopPropagation();
      if (closeEditPopup) {
        closeEditPopup();
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Container, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          multiple = _props.multiple,
          serviceConfig = _props.serviceConfig,
          onChange = _props.onChange,
          galleryConfig = _props.galleryConfig,
          closeEditPopup = _props.closeEditPopup;

      return _react2.default.createElement(
        Content,
        null,
        _react2.default.createElement(
          _antd.Tabs,
          { type: 'card' },
          _react2.default.createElement(
            TabPane,
            { tab: _react2.default.createElement(_reactIntl.FormattedMessage, { id: 'imgupload.tab1.title' }), key: '1' },
            _react2.default.createElement(_UploadImage2.default, {
              multiple: multiple,
              serviceConfig: serviceConfig,
              onChange: onChange,
              finishEdit: this.finishEdit
            })
          ),
          galleryConfig !== null && _react2.default.createElement(
            TabPane,
            { tab: _react2.default.createElement(_reactIntl.FormattedMessage, { id: 'imgupload.tab2.title' }), key: '2' },
            _react2.default.createElement(_DefaultImage2.default, {
              galleryConfig: galleryConfig,
              onChange: onChange
            })
          ),
          _react2.default.createElement(
            TabPane,
            { tab: _react2.default.createElement(_reactIntl.FormattedMessage, { id: 'imgupload.tab3.title' }), key: '3' },
            _react2.default.createElement(_UrlImage2.default, { onChange: onChange, closeEditPopup: closeEditPopup })
          )
        )
      );
    }
  }]);

  return Container;
}(_react.Component);

exports.default = Container;
