// @canner/image-upload/lib/DefaultImage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _templateObject = _taggedTemplateLiteral(["\n  background-image: url(", ");\n  background-position: 50%;\n  background-repeat: no-repeat;\n  background-size: cover;\n  margin: 0 auto;\n  height: 200px;\n  width: 100%;\n  cursor: pointer;\n"], ["\n  background-image: url(", ");\n  background-position: 50%;\n  background-repeat: no-repeat;\n  background-size: cover;\n  margin: 0 auto;\n  height: 200px;\n  width: 100%;\n  cursor: pointer;\n"]);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require("styled-components");

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _config = require("./config");

var _config2 = _interopRequireDefault(_config);

var _antd = require("antd");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var TabPane = _antd.Tabs.TabPane;

var DefaultImage = _styledComponents2.default.div(_templateObject, function (props) {
  return props.url;
});

var Gallery = (_temp2 = _class = function (_Component) {
  _inherits(Gallery, _Component);

  function Gallery() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Gallery);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Gallery.__proto__ || Object.getPrototypeOf(Gallery)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (data) {
      _this.props.onChange([data.url]);
    }, _this.generateContent = function (gallery) {
      var that = _this;
      var tabs = [];
      gallery.forEach(function (item, key) {
        tabs.push(_react2.default.createElement(
          TabPane,
          { key: key, tab: item.name },
          that.generateGallery(item.gallery)
        ));
      });

      return tabs;
    }, _this.generateGallery = function (gallery) {
      return _react2.default.createElement(
        _antd.Row,
        null,
        gallery.map(function (datum, i) {
          return _react2.default.createElement(
            _antd.Col,
            {
              style: { padding: "5px" },
              span: 8,
              key: i,
              onClick: function onClick() {
                return _this.handleClick(datum);
              }
            },
            _react2.default.createElement(DefaultImage, { url: datum.url })
          );
        })
      );
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Gallery, [{
    key: "render",
    value: function render() {
      var galleryConfig = this.props.galleryConfig;


      return _react2.default.createElement(
        _antd.Tabs,
        { tabPosition: "top" },
        this.generateContent(galleryConfig)
      );
    }
  }]);

  return Gallery;
}(_react.Component), _class.defaultProps = {
  galleryConfig: _config2.default
}, _temp2);
exports.default = Gallery;
