// @canner/slate-editor-html/lib/voidInlineRules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_OPT = {
  getSrc: function getSrc(node) {
    return node.data.get("src");
  }
};

function _default(Tag, inlineType) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : DEFAULT_OPT;
  return {
    deserialize: function deserialize(el) {
      if (inlineType && el.tagName && el.tagName.toLowerCase() === Tag) {
        var data = {};

        if (el.src) {
          data.src = el.src;
        }

        return {
          object: "inline",
          type: inlineType,
          data: data,
          isVoid: true
        };
      }
    },
    serialize: function serialize(obj) {
      if (obj.object == "inline" && obj.type === inlineType) {
        var src = options.getSrc(obj);
        var props = {
          src: src
        };
        return _react.default.createElement(Tag, props);
      }
    }
  };
}
