// @canner/slate-editor-html/lib/voidBlockRules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _react = _interopRequireDefault(require("react"));

var _lodash = _interopRequireDefault(require("lodash.mapvalues"));

var _slateIconShared = require("@canner/slate-icon-shared");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default(Tag, blockType) {
  var stylesAttr = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _slateIconShared.nodeAttrs;
  return {
    deserialize: function deserialize(el) {
      if (blockType && el.tagName && el.tagName.toLowerCase() === Tag) {
        var data = {};

        if (el.style.textAlign) {
          data.align = el.style.textAlign;
        }

        if (el.style.lineHeight) {
          data.lineHeight = el.style.lineHeight;
        }

        if (el.style.paddingLeft) {
          data.indent = el.style.paddingLeft;
        }

        if (Object.keys(data).length > 0) {
          return {
            object: "block",
            type: blockType,
            data: data,
            isVoid: true
          };
        }

        return {
          object: "block",
          type: blockType,
          isVoid: true
        };
      }
    },
    serialize: function serialize(obj) {
      if (obj.object == "block" && obj.type === blockType) {
        return _react.default.createElement(Tag, {
          style: (0, _lodash.default)(stylesAttr, function (val) {
            return val && val(obj);
          })
        });
      }
    }
  };
}
