// @canner/slate-editor-renderer/lib/components/imagePopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _slateHelperInlineAdddata = _interopRequireDefault(require("@canner/slate-helper-inline-adddata"));

var _antd = require("antd");

var _dec, _class;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var FormItem = _antd.Form.Item;

var _default = (_dec = _antd.Form.create(), _dec(_class =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ImageModal, _React$Component);

  function ImageModal() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, ImageModal);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(ImageModal)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleCancel", function () {
      _this.props.form.resetFields();

      _this.props.hidePopover();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleClickChange", function (visible) {
      if (!visible) _this.handleCancel();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleOk", function (e) {
      e.preventDefault();
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          change = _this$props.change,
          hidePopover = _this$props.hidePopover,
          form = _this$props.form,
          target = _this$props.target;
      change.select(target);
      form.validateFields(function (err, values) {
        if (!err) {
          var width = values.width,
              height = values.height;
          onChange(function (change) {
            change.call(_slateHelperInlineAdddata.default, {
              data: {
                width: Math.round(width),
                height: Math.round(height)
              }
            });
          });
          form.resetFields();
          hidePopover();
        }
      });
    });

    return _this;
  }

  _createClass(ImageModal, [{
    key: "render",
    value: function render() {
      var _this$props$form = this.props.form,
          getFieldDecorator = _this$props$form.getFieldDecorator,
          getFieldValue = _this$props$form.getFieldValue,
          setFieldsValue = _this$props$form.setFieldsValue;
      var _this$props2 = this.props,
          isEditing = _this$props2.isEditing,
          width = _this$props2.width,
          height = _this$props2.height,
          children = _this$props2.children;
      var content = React.createElement(_antd.Form, {
        horizontal: "true",
        layout: "inline"
      }, React.createElement(FormItem, {
        label: "Width:",
        hasFeedback: true
      }, getFieldDecorator("width", {
        rules: [{
          required: true,
          message: "Please enter a width"
        }],
        initialValue: width
      })(React.createElement(_antd.InputNumber, {
        onChange: function onChange(newWidth) {
          if (getFieldValue("lock")) {
            var _height = getFieldValue("height");

            var oldWidth = getFieldValue("width");
            var ratio = _height / oldWidth;
            setFieldsValue({
              height: (ratio * newWidth).toFixed(2)
            });
          }
        },
        onClick: function onClick(e) {
          return e.preventDefault();
        }
      }))), React.createElement(FormItem, {
        label: "Height:",
        hasFeedback: true
      }, getFieldDecorator("height", {
        rules: [{
          required: true,
          message: "Please enter a height"
        }],
        initialValue: height
      })(React.createElement(_antd.InputNumber, {
        onChange: function onChange(newHeight) {
          if (getFieldValue("lock")) {
            var _width = getFieldValue("width");

            var oldHeight = getFieldValue("height");
            var ratio = _width / oldHeight;
            setFieldsValue({
              width: (ratio * newHeight).toFixed(2)
            });
          }
        },
        onClick: function onClick(e) {
          return e.preventDefault();
        }
      }))), React.createElement(FormItem, null, getFieldDecorator("lock", {
        valuePropName: "checked",
        initialValue: true
      })(React.createElement(_antd.Checkbox, null, "Lock Original Ratio"))), React.createElement("div", null, React.createElement(_antd.Button, {
        key: "back",
        type: "ghost",
        onClick: this.handleCancel
      }, "Cancel"), " ", React.createElement(_antd.Button, {
        key: "submit",
        type: "primary",
        onClick: this.handleOk
      }, "Ok")));
      return React.createElement(_antd.Popover, {
        visible: isEditing,
        trigger: "click",
        title: "Set Image Size",
        content: content,
        onVisibleChange: this.handleClickChange,
        placement: "bottom"
      }, children);
    }
  }]);

  return ImageModal;
}(React.Component)) || _class);

exports.default = _default;
