// slate-trailing-block/dist/index.js
'use strict';

var Slate = require('slate');
var focusAtEnd = require('./focusAtEnd');

/**
 * Slate plugin to ensure a trailing block.
 * @param {Object} [opts] Options for the plugin
 * @param {String|Function} [opts.match='paragraph'] Match last block
 * @param {String} [opts.type] The type of the trailing block to insert
 * @return {Object}
 */

function TrailingBlock(opts) {
    opts = opts || {};
    opts.type = opts.type || 'paragraph';
    opts.match = opts.match || function (node) {
        return node.type === opts.type;
    };

    return {
        validateNode: function validateNode(node) {
            if (node.object !== 'document') {
                return undefined;
            }

            var lastNode = node.nodes.last();
            if (lastNode && opts.match(lastNode)) {
                return undefined;
            }

            var lastIndex = node.nodes.count();
            var block = Slate.Block.create({
                type: opts.type,
                nodes: [Slate.Text.create()]
            });

            return function (change) {
                return change.insertNodeByKey(node.key, lastIndex, block);
            };
        },

        changes: {
            focusAtEnd: focusAtEnd
        }
    };
}

module.exports = TrailingBlock;
